src/rpc/mining.cpp

```diff
@@ -572,24 +579,35 @@ UniValue getblocktemplate(const UniValue& params, bool fHelp)
             nTransactionsUpdatedLastLP = nTransactionsUpdatedLast;
         }
 
-        // Release the wallet and main lock while waiting
-        LEAVE_CRITICAL_SECTION(cs_main);
         {
-            checktxtime = boost::get_system_time() + boost::posix_time::minutes(1);
+            checktxtime = boost::get_system_time() + boost::posix_time::seconds(10);
 
             boost::unique_lock<boost::mutex> lock(csBestBlock);
             while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
             {
-                if (!cvBlockChange.timed_wait(lock, checktxtime))
-                {
-                    // Timeout: Check transactions for update
-                    if (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP)
-                        break;
-                    checktxtime += boost::posix_time::seconds(10);
+                // Before waiting, generate the coinbase for the block following the next
+                // block (since this is cpu-intensive), so that when next block arrives,
+                // we can quickly respond with the template for following block.
+                if (!next_coinbase_mtx && IsShieldedMinerAddress(minerAddress)) {
+                    next_coinbase_mtx = CoinbaseTransaction(Params(), CAmount{0}, minerAddress, chainActive.Tip()->nHeight + 2);
                 }
+                // Release the main lock while waiting
+                LEAVE_CRITICAL_SECTION(cs_main);
+                bool timedout = !cvBlockChange.timed_wait(lock, checktxtime);
+                ENTER_CRITICAL_SECTION(cs_main);
+
+                checktxtime += boost::posix_time::seconds(10);
+                // Timeout: Check transactions for update
+                if (timedout && mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP) {
+                    next_coinbase_mtx = std::nullopt;
+                    break;
+                }
+            }
+            if (next_coinbase_mtx && chainActive.Tip()->pprev->GetBlockHash() != hashWatchedChain) {
+                // reorg invalidates our precomputed coinbase tx.
+                next_coinbase_mtx = std::nullopt;
             }
```

Comment:
Now that I think about it, this is wrong. Recall that `hashWatchedChain` is the hash of what the RPC caller's considers to be the latest block (the tip). The caller wants us to wait until the hash of the tip block doesn't match this value (that's when we break out of the loop). _Normally_ (in the absence of a reorg), when a new block arrives, its parent is `hashWatchedChain`. So if we're inside this `if` block, it means a new block has arrived (else `next_coinbase_mtx` would be null), and it has some other parent than expected (the block that was the tip when we entered this RPC has been _replaced_). However, there's nothing in our precomputed coinbase that depends on the parent.

However, what if the height changes? That _is_ included in the coinbase. This is very unlikely, but if it happens, we do need to discard the precomputed coinbase. If we don't, we'll return an (empty) block with the wrong height in the coinbase, but the miner probably won't know that, and will mine an invalid block. Again, very unlikely (very rare), but this should be checking the height, not the hash. I'll fix it.